### PR TITLE
Rust constant caches

### PR DESCRIPTION
**Pull request by kddnewton**
Currently YJIT codegen invalidates code blocks whenever any constant is
changed. This commit changes that behavior to only invalidate code
blocks associated with the IDs that are being changed.

Changing this code required changing a bit about the way we're walking
through the ISEQs. In order to reuse the rb_iseq_each iterator, it was
necessary to make another callback that would be called whenever a
getconstant instruction was hit. This makes it nicer to interact with on
the rust side, but a little difficult to understand on the C side since
there are nested callbacks. Hopefully I've mitigated that enough with
comments.

### CODE REVIEW

**Review by maximecb — APPROVED** [review]

File: yjit/src/options.rs
Comment (XrXr) [inline yjit/src/options.rs:37]: I feel iffy about having a command line option to control invalidation behavior. Invalidation is core to the JIT and needs to be tested well while adding a flag to control its behavior makes the testing and the review more difficult due to the added code path. It's not going to run on CI so it's also prone to code rot.

File: vm_insnhelper.c
Comment (XrXr) [inline vm_insnhelper.c]:
```suggestion
    rb_vm_iseq_each_getconstant_id(iseq, cfp->pc - ISEQ_BODY(iseq)->iseq_encoded, vm_ic_compile_i, ic);
```

File: vm_insnhelper.c
Comment (XrXr) [inline vm_insnhelper.c]:
```suggestion
MJIT_STATIC void
```

I think this should fix the MJIT Windows CI. But, maybe it's irrelevant if you take the suggestion in https://github.com/Shopify/ruby/pull/267#pullrequestreview-932478569

**Review by XrXr — COMMENTED** [review]
I feel it would be better for the readability of both sides if we used `rb_iseq_each` from Rust land directly instead of making the C more generic. `vm_iseq_each_getconstant_id_i` is simple enough and I think in this case the reduced duplication is not worth having logic spread across multiple functions out of order in the C part. Right now to understand the Rust side you need to dig through the nested callback situation in vm_insnehelper.c.

I'm also a bit weary of the churn of touching the newly merged C side.

File: yjit/src/options.rs
Comment (maximecb) [reply yjit/src/options.rs:37]: That was my idea. I wanted to be able to get some stats with the old vs new invalidation behavior and it seemed more convenient. We could consider making this an environment variable instead (might be more convenient for Jean Boussier), and we can also plan to retire it once we've seen the data.

**Comment by maximecb** [comment]
> I feel it would be better for the readability of both sides if we used `rb_iseq_each` from Rust land directly instead of making the C more generic. `vm_iseq_each_getconstant_id_i` is simple enough and I think in this case the reduced duplication is not worth having logic spread across multiple functions out of order in the C part. Right now to understand the Rust side you need to dig through the nested callback situation in vm_insnehelper.c.

Yeah that part of the code is definitely not elegant. If it's possible to use `rb_iseq_each` and make the code a bit cleaner I agree that would be an improvement.

> I'm also a bit weary of the churn of touching the newly merged C side.

Would you prefer Kevin wait until the upstream C code is merged?

**Comment by XrXr** [comment]
> Would you prefer Kevin wait until the upstream C code is merged?

I think you mean after upstreaming Rust YJIT? The C side is already merged upstream. I think using `rb_iseq_each` makes it so we don't need to touch the C at all so it's fine to do now if indeed we don't need to touch the interpreter.

File: yjit/src/options.rs
Comment (XrXr) [reply yjit/src/options.rs:37]: Okay that's fair enough.

**Review by maximecb — APPROVED** [review]
🙏

**Pull request merged**
